store/pruning_block_store.go

```diff
@@ -0,0 +1,160 @@
+package store
+
+import (
+	"fmt"
+	"math"
+	"os"
+	"path"
+	"strconv"
+	"strings"
+
+	"github.com/loomnetwork/loomchain/log"
+	"github.com/tendermint/tendermint/blockchain"
+	dbm "github.com/tendermint/tendermint/libs/db"
+)
+
+var (
+	calcBlockMetaPrefix = []byte("H:")
+	blockStoreKey       = []byte("blockStore")
+)
+
+func calcBlockMetaKey(height int64) []byte {
+	return []byte(fmt.Sprintf("H:%v", height))
+}
+
+func calcBlockPartKey(height int64, partIndex int) []byte {
+	return []byte(fmt.Sprintf("P:%v:%v", height, partIndex))
+}
+
+func calcBlockCommitKey(height int64) []byte {
+	return []byte(fmt.Sprintf("C:%v", height))
+}
+
+func calcSeenCommitKey(height int64) []byte {
+	return []byte(fmt.Sprintf("SC:%v", height))
+}
+
+type PruningBlockStore struct {
+	blockStoreDB       dbm.DB
+	prunedBlockStoreDB dbm.DB
+	*blockchain.BlockStore
+	cfg       *BlockStoreConfig
+	batchSize uint64
+}
+
+func NewPruningBlockStore(blockStoreConfig *BlockStoreConfig) *PruningBlockStore {
+	blockStoreDB := dbm.NewDB("blockstore", "leveldb", path.Join(blockStoreConfig.ChainDataPath, "data"))
+	prunedBlockStoreDB := dbm.NewDB("pruned_blockstore", "leveldb", path.Join(blockStoreConfig.ChainDataPath, "data"))
+	return &PruningBlockStore{
+		blockStoreDB:       blockStoreDB,
+		prunedBlockStoreDB: prunedBlockStoreDB,
+		BlockStore:         blockchain.NewBlockStore(blockStoreDB),
+		cfg:                blockStoreConfig,
+	}
+}
+
+func (bs *PruningBlockStore) Close() {
+	bs.blockStoreDB.Close()
+	bs.prunedBlockStoreDB.Close()
+}
+
+func (bs *PruningBlockStore) Height() int64 {
+	return blockchain.LoadBlockStoreStateJSON(bs.blockStoreDB).Height
+}
+
+func (bs *PruningBlockStore) Prune() error {
+	latestHeight := bs.Height()
+	var targetHeight int64
+	targetHeight = latestHeight - bs.cfg.NumBlocksToRetain
+	graceBlocks := (bs.cfg.PruneGraceFactor / 100) * bs.cfg.NumBlocksToRetain
+	oldestHeight := int64(-1)
+	// Find the oldest block
+	it := bs.blockStoreDB.Iterator(calcBlockMetaPrefix, prefixRangeEnd(calcBlockMetaPrefix))
+	for ; it.Valid(); it.Next() {
+		oldestBlockMetaKey := it.Key()
+		oldestHeight = getHeightFromKey(oldestBlockMetaKey)
+		break
+	}
+	// Number of blocks to prune is less than grace blocks then skip pruning
+	if ((targetHeight - oldestHeight) + 1) < graceBlocks {
+		bs.blockStoreDB.Close()
+		bs.prunedBlockStoreDB.Close()
+		return nil
+	}
+	// If minimum height is greater than target height, there are no blocks to prune, skip pruning
+	if oldestHeight >= targetHeight {
+		bs.blockStoreDB.Close()
+		bs.prunedBlockStoreDB.Close()
+		return nil
+	}
+	batch := bs.prunedBlockStoreDB.NewBatch()
+	numHeight := int64(0)
+	for height := targetHeight; height <= latestHeight; height++ {
+		log.Info("Copying block at height", "height", height)
```

Comment:
This is going to be too spammy, should log the block range that will processed before the loop, and then the number of blocks actually processed at the end of the loop.